src/checkpoint/dispatch/vrt/serializer_registry.h

```diff
@@ -0,0 +1,150 @@
+//   contributors may be used to endorse or promote products derived from this
+//   software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact darma@sandia.gov
+//
+// *****************************************************************************
+//@HEADER
+*/
+
+#if !defined INCLUDED_CHECKPOINT_DISPATCH_VRT_SERIALIZER_REGISTRY_H
+#define INCLUDED_CHECKPOINT_DISPATCH_VRT_SERIALIZER_REGISTRY_H
+
+#include "checkpoint/common.h"
+#include "checkpoint/dispatch/vrt/registry_common.h"
+
+#include <functional>
+#include <vector>
+#include <tuple>
+
+namespace checkpoint { namespace dispatch { namespace vrt {
+namespace serializer_registry {
+
+template <typename ObjT>
+using RegistryType = std::vector<
+  std::tuple<
+    TypeIdx,                             /**< This entry type idx */
+    TypeIdx,                             /**< The base class type idx */
+    std::function<void(void*, ObjT&)>    /**< Type-erased serialize dispatch */
+  >
+>;
+
+template <typename ObjT, typename SerializerT>
+struct Registrar {
+  Registrar();
+  TypeIdx index;
+};
+
+template <typename ObjT, typename SerializerT>
+struct Type {
+  static TypeIdx const idx;
+};
+
+template <typename ObjT>
+inline RegistryType<ObjT>& getRegistry() {
+  static RegistryType<ObjT> reg;
+  return reg;
+}
+
+template <typename ObjT, typename SerializerT>
+inline TypeIdx makeObjIdx() {
+  return Type<ObjT, SerializerT>::idx;
+}
+
+template <typename ObjT, typename SerializerT>
+Registrar<ObjT, SerializerT>::Registrar() {
+  auto& reg = getRegistry<ObjT>();
+  index = static_cast<TypeIdx>(reg.size());
+
+  debug_checkpoint(
+    "serializer registrar: %zu, ObjT=%s SerializerT=%s\n",
+    reg.size(),
+    typeid(ObjT).name(),
+    typeid(SerializerT).name()
+  );
+
+  reg.emplace_back(
+    index,
+    no_type_idx, // Set later when linkDerivedToBase is called
+    [=](void* s, ObjT& obj) {
+      auto& ser = *reinterpret_cast<SerializerT*>(s);
+      // Disable virtual serializer dispatch because we are already in a
+      // virtualSerialize and otherwise we will recurse indefinitely
+      ser.setVirtualDisabled(true);
```

Comment:
Does this need to be a stack where we save the value before pushing 'true' here, and restore it when we come back out? Consider recursive nested objects.

Reply:
It gets immediately unset when it triggers the non-virtual serialize in the dispatcher. I have a test with nesting to ensure this works properly.

Reply:
Initially, I thought a stack might be needed also. But I believe a stack isn't needed because the state is only valid for exactly one call depth and then unset.

Reply:
Ok, I understand that now. It would be helpful to clarify in the comment that it immediately gets unset on the other side of the fence.

Reply:
And probably have a matching comment where that happens, pointing here.